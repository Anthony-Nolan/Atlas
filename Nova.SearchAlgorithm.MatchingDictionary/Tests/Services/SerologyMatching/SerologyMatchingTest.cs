using System.Collections.Generic;
using System.Linq;
using ApprovalTests;
using ApprovalTests.Reporters;
using Nova.SearchAlgorithm.MatchingDictionary.Models.HLATypes;
using Nova.SearchAlgorithm.MatchingDictionary.Models.MatchingTypes;
using NUnit.Framework;

namespace Nova.SearchAlgorithm.MatchingDictionary.Tests.Services.SerologyMatching
{
    [TestFixtureSource(typeof(SerologyMatchingTestFixtureArgs))]
    [UseReporter(typeof(NUnitReporter))]
    public class SerologyMatchingTest : MatchedOnTestBase<IMatchingSerology>
    {
        public SerologyMatchingTest(IEnumerable<IMatchingSerology> matchingSerology) : base(matchingSerology)
        {
        }

        [Test]
        public void ServiceReturnsAllExpectedSerology()
        {
            var str = string.Join("\r\n", MatchingTypes
                .OrderBy(s => s.HlaType.MatchLocus)
                .ThenBy(s => int.Parse(s.HlaType.Name))
                .Select(s => $"{s.HlaType.MatchLocus}\t{s.HlaType.Name}")
                .ToList());
            Approvals.Verify(str);
        }

        [Test]
        public void BroadMatchingSerologiesAreCorrect()
        {
            var broadWhereSplitHasAssociated = new SerologyToSerology
            (
                new Serology("A", "9", Subtype.Broad),
                new Serology("A", "9", Subtype.Broad),
                new List<Serology>
                {
                    new Serology("A", "9", Subtype.Broad),
                    new Serology ("A", "23", Subtype.Split),
                    new Serology ("A", "24", Subtype.Split),
                    new Serology ("A", "2403", Subtype.Associated)
                }
            );

            var broadWhereSplitHasNoAssociated = new SerologyToSerology
            (
                new Serology("DQ", "1", Subtype.Broad),
                new Serology("DQ", "1", Subtype.Broad),
                new List<Serology>
                {
                    new Serology("DQ", "1", Subtype.Broad),
                    new Serology ("DQ", "5", Subtype.Split),
                    new Serology ("DQ", "6", Subtype.Split)
                }
            );

            var broadHasSplitAndAssociated = new SerologyToSerology
            (
                new Serology("B", "21", Subtype.Broad),
                new Serology("B", "21", Subtype.Broad),
                new List<Serology>
                {
                    new Serology("B", "21", Subtype.Broad),
                    new Serology ("B", "4005", Subtype.Associated),
                    new Serology ("B", "49", Subtype.Split),
                    new Serology ("B", "50", Subtype.Split)
                }
            );

            Assert.AreEqual(broadWhereSplitHasAssociated, GetSingleMatchingType("A", "9"));
            Assert.AreEqual(broadWhereSplitHasNoAssociated, GetSingleMatchingType("DQB1", "1"));
            Assert.AreEqual(broadHasSplitAndAssociated, GetSingleMatchingType("B", "21"));
        }

        [Test]
        public void SplitMatchingSerologiesAreCorrect()
        {
            var splitHasAssociated = new SerologyToSerology
            (
                new Serology("B", "51", Subtype.Split),
                new Serology("B", "51", Subtype.Split),
                new List<Serology>
                {
                    new Serology("B", "51", Subtype.Split),
                    new Serology ("B", "5", Subtype.Broad),
                    new Serology ("B", "5102", Subtype.Associated),
                    new Serology ("B", "5103", Subtype.Associated)
                }
            );

            var splitNoAssociated = new SerologyToSerology
            (
                new Serology("Cw", "10", Subtype.Split),
                new Serology("Cw", "10", Subtype.Split),
                new List<Serology>
                {
                    new Serology("Cw", "10", Subtype.Split),
                    new Serology ("Cw", "3", Subtype.Broad)
                }
            );

            Assert.AreEqual(splitHasAssociated, GetSingleMatchingType("B", "51"));
            Assert.AreEqual(splitNoAssociated, GetSingleMatchingType("C", "10"));
        }

        [Test]
        public void AssociatedMatchingSerologiesAreCorrect()
        {
            var associatedWithSplit = new SerologyToSerology
            (
                new Serology("B", "3902", Subtype.Associated),
                new Serology("B", "3902", Subtype.Associated),
                new List<Serology>
                {
                    new Serology("B", "3902", Subtype.Associated),
                    new Serology ("B", "39", Subtype.Split),
                    new Serology ("B", "16", Subtype.Broad)
                }
            );

            var associatedWithBroad = new SerologyToSerology
            (
                new Serology("B", "4005", Subtype.Associated),
                new Serology("B", "4005", Subtype.Associated),
                new List<Serology>
                {
                    new Serology("B", "4005", Subtype.Associated),
                    new Serology ("B", "21", Subtype.Broad)
                }
            );

            var associatedWithNotSplit = new SerologyToSerology
            (
                new Serology("DR", "103", Subtype.Associated),
                new Serology("DR", "103", Subtype.Associated),
                new List<Serology>
                {
                    new Serology("DR", "103", Subtype.Associated),
                    new Serology ("DR", "1", Subtype.NotSplit)
                }
            );

            Assert.AreEqual(associatedWithSplit, GetSingleMatchingType("B", "3902"));
            Assert.AreEqual(associatedWithBroad, GetSingleMatchingType("B", "4005"));
            Assert.AreEqual(associatedWithNotSplit, GetSingleMatchingType("DRB1", "103"));
        }

        [Test]
        public void NotSplitMatchingSerologiesAreCorrect()
        {
            var notSplitWithAssociated = new SerologyToSerology
            (
                new Serology("A", "2", Subtype.NotSplit),
                new Serology("A", "2", Subtype.NotSplit),
                new List<Serology>
                {
                    new Serology("A", "2", Subtype.NotSplit),
                    new Serology ("A", "203", Subtype.Associated),
                    new Serology ("A", "210", Subtype.Associated)
                }
            );

            var notSplitNoAssociatedA = new SerologyToSerology
            (
                new Serology("A", "1", Subtype.NotSplit),
                new Serology("A", "1", Subtype.NotSplit),
                new List<Serology> { new Serology("A", "1", Subtype.NotSplit) }
            );

            var notSplitNoAssociatedB = new SerologyToSerology
            (
                new Serology("B", "13", Subtype.NotSplit),
                new Serology("B", "13", Subtype.NotSplit),
                new List<Serology> { new Serology("B", "13", Subtype.NotSplit) }
            );

            var notSplitNoAssociatedC = new SerologyToSerology
            (
                new Serology("Cw", "8", Subtype.NotSplit),
                new Serology("Cw", "8", Subtype.NotSplit),
                new List<Serology> { new Serology("Cw", "8", Subtype.NotSplit) }
            );

            var notSplitNoAssociatedDq = new SerologyToSerology
            (
                new Serology("DQ", "4", Subtype.NotSplit),
                new Serology("DQ", "4", Subtype.NotSplit),
                new List<Serology> { new Serology("DQ", "4", Subtype.NotSplit) }
            );

            var notSplitNoAssociatedDr = new SerologyToSerology
            (
                new Serology("DR", "9", Subtype.NotSplit),
                new Serology("DR", "9", Subtype.NotSplit),
                new List<Serology> { new Serology("DR", "9", Subtype.NotSplit) }
            );

            Assert.AreEqual(notSplitWithAssociated, GetSingleMatchingType("A", "2"));
            Assert.AreEqual(notSplitNoAssociatedA, GetSingleMatchingType("A", "1"));
            Assert.AreEqual(notSplitNoAssociatedB, GetSingleMatchingType("B", "13"));
            Assert.AreEqual(notSplitNoAssociatedC, GetSingleMatchingType("C", "8"));
            Assert.AreEqual(notSplitNoAssociatedDq, GetSingleMatchingType("DQB1", "4"));
            Assert.AreEqual(notSplitNoAssociatedDr, GetSingleMatchingType("DRB1", "9"));
        }

        [Test]
        public void DeletedMatchingSerologiesAreCorrect()
        {
            var deletedSerology = new SerologyToSerology
            (
                new Serology("Cw", "11", Subtype.NotSplit, true),
                new Serology("Cw", "1", Subtype.NotSplit),
                new List<Serology>
                {
                    new Serology("Cw", "11", Subtype.NotSplit, true),
                    new Serology("Cw", "1", Subtype.NotSplit)
                }
            );

            Assert.AreEqual(deletedSerology, GetSingleMatchingType("C", "11"));
        }

        [Test]
        public void MatchingSerologyOnlyContainsValidRelationships()
        {
            var groupBySubtype = MatchingTypes
                .Where(m => !m.HlaType.IsDeleted)
                .Select(m => new
                {
                    MatchedType = (Serology)m.HlaType,
                    SubtypeCounts = m.MatchingSerologies
                        .Where(s => !s.Equals(m.HlaType))
                        .GroupBy(s => s.Subtype)
                        .Select(s => new { Subtype = s.Key, Count = s.Count() })
                }).ToList();

            var broads = groupBySubtype.Where(s => s.MatchedType.Subtype == Subtype.Broad).ToList();
            var splits = groupBySubtype.Where(s => s.MatchedType.Subtype == Subtype.Split).ToList();
            var associated = groupBySubtype.Where(s => s.MatchedType.Subtype == Subtype.Associated).ToList();
            var notSplits = groupBySubtype.Where(s => s.MatchedType.Subtype == Subtype.NotSplit).ToList();

            // Matching list should not contain the subtype of the Matched type
            Assert.IsFalse(
                groupBySubtype.Any(s => s.SubtypeCounts.Any(sc => sc.Subtype == s.MatchedType.Subtype)));

            // Broads cannot be matched to NotSplit, and must have at least two Splits
            Assert.IsFalse(
                broads.Any(b => b.SubtypeCounts.Any(sc => sc.Subtype == Subtype.NotSplit)));
            Assert.IsFalse(
                broads.Any(b => b.SubtypeCounts.Single(sc => sc.Subtype == Subtype.Split).Count < 2));

            // Splits cannot be matched to NotSplit, and must have one Broad
            Assert.IsFalse(
                splits.Any(s => s.SubtypeCounts.Any(sc => sc.Subtype == Subtype.NotSplit)));
            Assert.IsFalse(
                splits.Any(s => s.SubtypeCounts.Single(sc => sc.Subtype == Subtype.Broad).Count != 1));

            // Associated can only have:
            //      * 1 x NotSplit, or;
            //      * 1 x Broad, or;
            //      * 1 x Split and 1 x Broad
            Assert.IsFalse(
                associated.Any(a => a.SubtypeCounts.Any(sc => sc.Count != 1)));
            Assert.IsFalse(
                associated.Any(a =>
                    a.SubtypeCounts.Any(sc => sc.Subtype == Subtype.NotSplit)
                    && a.SubtypeCounts.Any(sc => sc.Subtype != Subtype.NotSplit)));
            Assert.IsFalse(
                associated
                .Where(a => a.SubtypeCounts.Any(sc => sc.Subtype == Subtype.Split))
                .Any(a => a.SubtypeCounts.All(sc => sc.Subtype != Subtype.Broad)));

            // NotSplits can only be matched to Associated
            Assert.IsFalse(
                notSplits.Any(n => n.SubtypeCounts.Any(sc => sc.Subtype != Subtype.Associated)));
        }
    }
}
