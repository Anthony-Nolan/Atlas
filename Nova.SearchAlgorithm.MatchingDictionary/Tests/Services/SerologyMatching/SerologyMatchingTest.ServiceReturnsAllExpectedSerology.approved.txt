A	1
A	2
A	3
A	9
A	10
A	11
A	19
A	23
A	24
A	25
A	26
A	28
A	29
A	30
A	31
A	32
A	33
A	34
A	36
A	43
A	66
A	68
A	69
A	74
A	80
A	203
A	210
A	2403
B	5
B	7
B	8
B	12
B	13
B	14
B	15
B	16
B	17
B	18
B	21
B	22
B	27
B	35
B	37
B	38
B	39
B	40
B	41
B	42
B	44
B	45
B	46
B	47
B	48
B	49
B	50
B	51
B	52
B	53
B	54
B	55
B	56
B	57
B	58
B	59
B	60
B	61
B	62
B	63
B	64
B	65
B	67
B	70
B	71
B	72
B	73
B	75
B	76
B	77
B	78
B	81
B	82
B	703
B	2708
B	3901
B	3902
B	4005
B	5102
B	5103
C	1
C	2
C	3
C	4
C	5
C	6
C	7
C	8
C	9
C	10
C	11
DQB1	1
DQB1	2
DQB1	3
DQB1	4
DQB1	5
DQB1	6
DQB1	7
DQB1	8
DQB1	9
DRB1	1
DRB1	2
DRB1	3
DRB1	4
DRB1	5
DRB1	6
DRB1	7
DRB1	8
DRB1	9
DRB1	10
DRB1	11
DRB1	12
DRB1	13
DRB1	14
DRB1	15
DRB1	16
DRB1	17
DRB1	18
DRB1	103
DRB1	1403
DRB1	1404